### PR TITLE
:rotating_light: :rotating_light: :rotating_light: [Breaking change] Deformable DETR intermediate representations

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

- Fixes naturally the `object-detection` pipeline.
- Moves from `[n_decoders, batch_size, ...]` to `[batch_size,
  n_decoders, ...]` instead.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/deformable_detr/modeling_deformable_detr.py [hunk 1]
```diff
@@ -1394,8 +1394,9 @@ def custom_forward(*inputs):
                 if encoder_hidden_states is not None:
                     all_cross_attentions += (layer_outputs[2],)
 
-        intermediate = torch.stack(intermediate)
-        intermediate_reference_points = torch.stack(intermediate_reference_points)
+        # XXX: Keep batch_size as first dimension
```
Comment (sgugger) [inline src/transformers/models/deformable_detr/modeling_deformable_detr.py:1397]:
```suggestion
        # Keep batch_size as first dimension
```

File: src/transformers/models/deformable_detr/modeling_deformable_detr.py [hunk 2]
```diff
@@ -1935,11 +1936,12 @@ def forward(
             outputs_coord = outputs_coord_logits.sigmoid()
             outputs_classes.append(outputs_class)
             outputs_coords.append(outputs_coord)
-        outputs_class = torch.stack(outputs_classes)
-        outputs_coord = torch.stack(outputs_coords)
+        # XXX: Keep batch_size as first dimension
```
Comment (sgugger) [inline src/transformers/models/deformable_detr/modeling_deformable_detr.py:1939]:
```suggestion
        # Keep batch_size as first dimension
```

**Review by sgugger — APPROVED** [review]
LGTM, thanks for fixing! Can you just put three :rotating_light: in the title so that we can easily identify this PR as we list breaking changes when doing next release?

**Review by NielsRogge — APPROVED** [review]
Thanks a lot for working on this.

Could you verify the slow integration tests are passing please?

**Comment by Narsil** [comment]
```
==================================================================================================== 73 passed, 8 skipped, 49 warnings in 98.91s (0:01:38) =====================================================================================================

```
Is that OK (I think the skipped are just unsupported features like resize_tokens)

**Comment by Narsil** [comment]
@sgugger Since this is breaking, I'll let you merge when it's more convenient.

**Comment by sgugger** [comment]
Let's go :-)

**Pull request merged**